#751 fix - Replaces decimals from backend with decimals from blockchain

Agupane opened the pull request:
Closes #751

## Description
- As described in the ticket, we found that some decimals from the backend where wrong, so we decided to fix this taking the decimals from the blockchain.

gnosis-info commented:
Travis automatic deployment:
 https://pr755--safereact.review.gnosisdev.com/app

mikheevm commented:
I wouldn't merge this until we figure out where incorrect values were coming from.

fernandomg commented:
> I wouldn't merge this until we figure out where incorrect values were coming from.
> 
> Also
> 
> > we found that some decimals from the backend where wrong
> 
> there's no evidence of that in ticket's discussion

Agree.

gnosis-info commented:
Travis automatic deployment:
 https://pr755--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr755--safereact.review.gnosisdev.com/app

Agupane commented:
Ok after doing some research for a while with @fernandomg, we found that the problem was when we `JSON.stringify` the token object, to store it in localStorage. It happens that `truffle-contract `returns a `BN` (bn.js) instance for numeric values. BN handles values internally as hex values but without the leading 0x, so when it's stringified the toJSON method is called, and it returns the hex value without the leading 0x giving us a 12 instead of an 18 (0x12).
When that value was recovered from the localStorage, it was treated as a valid string representation of the decimal value.

In order to fix this, we converted the BigNumber object to a number (using toNumber method) before calling `makeToken`. Also, for the tokens that are already stored in the user's localStorage, we removed them from then list checking the type of the object.

```
> decimals = web3Utils.toBN('18')
<BN: 12>
> JSON.stringify({ decimals })
'{"decimals":"12"}'
> JSON.parse(JSON.stringify({ decimals }))
{ decimals: '12' }
> new bn(JSON.parse(JSON.stringify({ decimals })).decimals, 16).toNumber()
18
```

Agupane commented:
@francovenica in order to test this, you need to look for a token that's not present on the backend, like `0xEc0B6AFb3F9a609CEEd67e2CA551A4c573fd45F7 `. You can see that on both localStorage and redux, the decimals for that tokens is was 12. Now should be 18, must also update data from localStorage's with 12 decimals.

mikheevm commented on src/logic/tokens/store/actions/loadActiveTokens.js:
```diff
@@ -12,7 +12,9 @@ const loadActiveTokens = () => async (dispatch: ReduxDispatch<GlobalState>) => {
     const tokenRecordsList: List<Token> = List(
-      Object.values(tokens).map((token: TokenProps): Token => makeToken(token)),
+      Object.values(tokens)
+        .filter((t) => typeof t.decimals !== 'string')
```
could you maybe add a comment there explaining why we're doing the filter (link to your PR comment with explanation would be enough)

Agupane replied:
Yeah good idea, we will forget about this in the future, will add it

mikheevm commented:
@fernandomg @Agupane 
![image](https://user-images.githubusercontent.com/16622558/79143168-cd0a6680-7dcd-11ea-86aa-754cc97e8407.png)

gnosis-info commented:
Travis automatic deployment:
 https://pr755--safereact.review.gnosisdev.com/app

francovenica commented:
@fernandomg @Agupane 
I've tried by using the safe where I reported the issue initially. 
Went to the Transaction tab and then switched to the 2nd page (25 rows per page) and the site crashed throwing a ton of errors.
Can you tell me if those are related to this ticket?

The safe:
https://pr755--safereact.review.gnosisdev.com/app/#/safes/0xc3214e72A53695faC3A124C92FcA3f161D193299/transactions

Snapshot:
![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/b82aa58d-9078-4a8b-a3f0-358f802b6b46)

mikheevm commented:
@francovenica this is related to the safe apps @nicosampler

francovenica commented:
@mikheevm gotcha. Thanks. I'll ignore it for this ticket

I've created a new safe and transferred some funds to it: BAT, REP, ZRX
They now show the right values:

The safe: 
https://pr755--safereact.review.gnosisdev.com/app/#/safes/0x4816Fae09FdFaF6927E785EFe68E5D68f5740840/transactions

Snapshot:
![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/57796b00-b755-4615-aedf-5b5859318132)

Also as Agu Pane told me I've checked in the redux and they now show 18 decimals:
![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/5987cae7-4ab4-4881-afcb-ece2d4406bef)
![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/c3b2aa9f-1688-47a5-8059-5aa79acc462f)
![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/4413eae3-7382-4dd0-907a-7a845cea2783)

Looks good to me
